keps/sig-scheduling/4671-gang-scheduling/README.md

```diff
@@ -0,0 +1,983 @@
+// multiple dense indexes (PodSubGroups) and which can optionally be replicated in a variable
+// number of identical copies.
+type PodGroup struct {
+    Name *string
+
+    // Number of identical instances of PodGroup that are part of the Workload.
+    // Defaults to 1.
+    Replicas int
+
+    // Policy defines the configuration of the PodGroup to enable different
+    // scheduling policies.
+    Policy PodGroupPolicy
+}
+
+// PodGroupPolicy defines scheduling configuration of a PodGroup.
+type PodGroupPolicy struct {
+    // Exactly one of the policies should be set.
+
+    // Default scheduling policy (default Kubernetes behavior).
+    Default *DefaultSchedulingPolicy
+
+    // Gang scheduling policy (all-or-nothing scheduling semantics)
+    Gang *GangSchedulingPolicy
+}
+
+// DefaultSchedulingPolicy represents default scheduling behavior.
+type DefaultSchedulingPolicy struct {
+    // For now this is effectively just a marker type.
+}
+
+// GangSchedulingPolicy represents options for how gang scheduling of one
+// PodGroup should be handled.
+type GangSchedulingPolicy struct {
+    MinCount *int
+}
+
+type WorkloadStatus struct {
+  // Necessary status fields TBD.
+}
+```
+
+The individual `PodGroups` and `PodGroup` replicas are treated as independent gangs. As an example, if one of
+the groups can be scheduled and the other can't be - this is exactly what will happen. If the underlying
+user intention was to have either both of them or none of them running, they should form a single group and
+not be split into two. A `LeaderWorkerSet` is a good example of it, where a single `PodGroup` replica consists
+of a single leader and `N` workers and that forms a scheduling (and runtime unit), but workload as a whole
+may consist of a number of such replicas.
+
+### Scheduler Changes
+
+The kube-scheduler will be watching for `Workload` objects (using informers) and will use them to map pods
+to and from their `Workload` objects.
+
+In the initial implementation, we expect users to create the `Workload` objects. In the next steps controllers
+will be updated to create an appropriate `Workload` objects themselves whenever they can appropriately infer
+the intention from the desired state.
+Note that given scheduling options are stored in the `Workload` object, pods linked to the `Workload`
+object will not be scheduled until this `Workload` object is created and observed by the kube-scheduler.
+
+The north star vision for gang scheduling implementation should satisfy the following requirements:
+
+1. Ensure that pods being part of a gang are not bound if all pods belonging to it can't be scheduled.
+2. Provide the "optimal enough" placement by considering all pods from a gang together.
+3. Avoid deadlock scenario when multiple workloads are being scheduled at the same time by kube-scheduler.
+4. Avoid deadlock scenario when multiple workloads are being scheduled at the same time by different
```

Comment:
Any particular reason you have this element mentioned twice? Why can't you just write this as:
```suggestion
4. Avoid deadlock scenario when multiple workloads are being scheduled at the same time by any scheduler (kube-scheduler, or third-party provided scheduler). 
```